### PR TITLE
Support redis user password auth and tls encryption

### PR DESCRIPTION
**Pull request by skjnldsv**
``` php
'memcache.locking' => '\OC\Memcache\Redis',
'memcache.distributed' => '\OC\Memcache\Redis',
'memcache.local' =>'\OC\Memcache\Redis' ,
'redis' => [
	  'host' => 'tls://127.0.0.1',
	  'port' => 6379,
	  'user' => 'nextcloud',
	  'password' => 'password',
	  'ssl_context' => [
		    'local_cert' => '/certs/redis.crt',
		    'local_pk' => '/certs/redis.key',
		    'cafile' => '/certs/ca.crt',
		    'verify_peer_name' => false
	  ]
]
```
(From https://redis.io/topics/encryption)

1. Download redis source https://github.com/redis/redis/archive/refs/tags/6.2.4.tar.gz
2. Generate certificates from `./utils/gen-test-certs.sh`
3. Create a users.acl config file containing `user nextcloud on +@all ~* >password`
3. Start redis server
  ```bash
  redis-server --tls-port 6379 --port 0 \
    --tls-cert-file ./certs/redis.crt \
    --tls-key-file ./certs/redis.key \
    --tls-ca-cert-file ./certs/ca.crt \
    --tls-auth-clients optional \
    --aclfile users.acl \
    --protected-mode no
  ```
4. Setup nextcloud with above config
5. Profit

___
## Questions
- [x] How shall we handle the support and throw of old phpredis module version? Is there a way to say `phpredis > 5.x.x` for Nextcloud 23 ?

### CODE REVIEW

**Comment by kesselb** [comment]
To fix Psalm: https://github.com/nextcloud/server/pull/27888

**Comment by kesselb** [comment]
```
    /**
     * Authenticate the connection using a password.
     * Warning: The password is sent in plain-text over the network.
     *
     * @param string|string[] $password
     *
     * @return bool TRUE if the connection is authenticated, FALSE otherwise
     *
     * @link    https://redis.io/commands/auth
     * @example $redis->auth('foobared');
     */
    public function auth($password)
    {
    }
```

Please update build/stubs/redis.php with the above to fix Psalm. That's faster than updating the stubs (because they are using php8 attributes now).

**Comment by skjnldsv** [comment]
```
PHP Warning:  Redis::connect() expects at most 6 parameters, 7 given in /drone/src/lib/private/RedisFactory.php on line 112
24 | An unhandled exception has been thrown:
25 | RedisException: Redis server went away in /drone/src/lib/private/Memcache/Redis.php:55

```

How shall we support this? Is it because we're running an old phpredis version I guess? We're using 7.3 for those tests. 
`image: ghcr.io/nextcloud/continuous-integration-integration-php7.3:integration-php7.3-2`, Latest phpredis requires PHP 7.0.0 or newer. So we should be good? :thinking: 
Maybe we need to update the  docker images?

In any case, let me update the docker images

**Comment by skjnldsv** [comment]
@LukasReschke how do we push the docker--ci tags to the `ghcr.io/nextcloud` repo?

**Comment by juliushaertl** [comment]
> How shall we support this? Is it because we're running an old phpredis version I guess? We're using 7.3 for those tests.

Depends a bit on when this was introduced I'd say since I could imagine not all distribution packages are using the latest phpredis version. We could basically check for the used extension version with https://www.php.net/manual/en/function.phpversion.php

**Comment by PVince81** [comment]
@skjnldsv as discussed in chat, no SSL support was added for RedisCluster for now, correct ?

From what I remember it would require copying all the config values from the "ssl_context" attribute into the main config object.

**Comment by skjnldsv** [comment]
Yes, I ignored clustering for now

**Comment by PVince81** [comment]
I've tested this PR and it worked for me locally with the instructions.

@skjnldsv I think we talked about the possibility of adding/updating unit tests for the RedisFactory, is this still to do ?

**Comment by skjnldsv** [comment]
> @skjnldsv I think we talked about the possibility of adding/updating unit tests for the RedisFactory, is this still to do ?

I'm lacking the knowledge and time for that, wanna focus on it?

**Comment by PVince81** [comment]
> > @skjnldsv I think we talked about the possibility of adding/updating unit tests for the RedisFactory, is this still to do ?
> 
> I'm lacking the knowledge and time for that, wanna focus on it?

I had a look and it would require some extra gymnastics (refactoring, injection) to be able to mock `Redis` and `RedisCluster` and considering the low complexity it's probably not worth the effort.

**Comment by PVince81** [comment]
> > How shall we support this? Is it because we're running an old phpredis version I guess? We're using 7.3 for those tests.
> 
> Depends a bit on when this was introduced I'd say since I could imagine not all distribution packages are using the latest phpredis version. We could basically check for the used extension version with https://www.php.net/manual/en/function.phpversion.php

Sounds like a good plan: if extra "ssl_context" options are provided but not supported by the current redis module version, throw an exception

**Comment by skjnldsv** [comment]
Done, I upgraded and refactored everything for cleaner code.
Please review again @PVince81 and @juliushaertl

**Review by PVince81 — APPROVED** [review]
:+1: code looks fine, I didn't retest though.

cool that you were able to also add the bits for RedisCluster

File: config/config.sample.php [hunk 1]
```diff
@@ -1222,13 +1222,25 @@
...
  */
 'redis' => [
 	'host' => 'localhost', // can also be a unix domain socket: '/tmp/redis.sock'
 	'port' => 6379,
 	'timeout' => 0.0,
+	'read_timeout' => 0.0,
+	'user' =>  '', // Optional, if not defined no password will be used.
 	'password' => '', // Optional, if not defined no password will be used.
 	'dbindex' => 0, // Optional, if undefined SELECT will not run and will use Redis Server's default DB Index.
+	// If redis is encrypted, provide certificates
```
[the remark this answers is not in the document]
Comment (skjnldsv) [reply config/config.sample.php:1237]: In-transit encryption. We should indeed be more clear :)

File: lib/private/RedisFactory.php [hunk 2]
```diff
@@ -43,54 +46,82 @@ public function __construct(SystemConfig $config) {
 	}
 
 	private function create() {
-		if ($config = $this->config->getValue('redis.cluster', [])) {
+		$isCluster = !empty($this->config->getValue('redis.cluster', []));
+		$config = $this->config->getValue('redis', []);
```
[the remark this answers is not in the document]
Comment (skjnldsv) [reply lib/private/RedisFactory.php:50]: Yeah, I think it's also a bit worst in terms of code readability?
I like having the base config on top.

[the remark this answers is not in the document]
Comment (skjnldsv) [reply lib/private/RedisFactory.php:50]: But I could do something like this?
```php
	private function create() {
		$isCluster = !empty($this->config->getValue('redis.cluster', []));
		$config = $isCluster
			? $this->config->getValue('redis.cluster', []);
			: $this->config->getValue('redis', []);

		// Init cluster config if any
		if ($isCluster && !class_exists('RedisCluster')) {
			throw new \Exception('Redis Cluster support is not available');
		}

```

**Comment by skjnldsv** [comment]
Okay, I went down the rabbit hole

Tests setups NO TLS
```php
    'memcache.locking' => '\OC\Memcache\Redis',
    'memcache.distributed' => '\OC\Memcache\Redis',
    'memcache.local' =>'\OC\Memcache\Redis' ,
    'redis.cluster' => array(
      'seeds' => [
        'redis-node-0:6379',
        'redis-node-1:6379',
        'redis-node-2:6379',
        'redis-node-3:6379',
        'redis-node-4:6379',
        'redis-node-5:6379',
      ],
      'failover_mode' => \RedisCluster::FAILOVER_ERROR,
      'password' => 'password' # optional if you're using the protected config
    ),
```

<details>
<summary>Docker-compose unprotected cluster</summary>

```yml
    redis-node-0:
        image: docker.io/bitnami/redis-cluster
        environment:
          - ALLOW_EMPTY_PASSWORD=yes
          - REDIS_NODES=redis-node-0 redis-node-1 redis-node-2 redis-node-3 redis-node-4 redis-node-5
        networks:
            - nextcloud

    redis-node-1:
        image: docker.io/bitnami/redis-cluster
        environment:
          - ALLOW_EMPTY_PASSWORD=yes
          - REDIS_NODES=redis-node-0 redis-node-1 redis-node-2 redis-node-3 redis-node-4 redis-node-5
        networks:
            - nextcloud

    redis-node-2:
        image: docker.io/bitnami/redis-cluster
        environment:
          - ALLOW_EMPTY_PASSWORD=yes
          - REDIS_NODES=redis-node-0 redis-node-1 redis-node-2 redis-node-3 redis-node-4 redis-node-5
        networks:
            - nextcloud

    redis-node-3:
        image: docker.io/bitnami/redis-cluster
        environment:
          - ALLOW_EMPTY_PASSWORD=yes
          - REDIS_NODES=redis-node-0 redis-node-1 redis-node-2 redis-node-3 redis-node-4 redis-node-5
        networks:
            - nextcloud

    redis-node-4:
        image: docker.io/bitnami/redis-cluster
        environment:
          - ALLOW_EMPTY_PASSWORD=yes
          - REDIS_NODES=redis-node-0 redis-node-1 redis-node-2 redis-node-3 redis-node-4 redis-node-5
        networks:
            - nextcloud

    redis-node-5:
        image: docker.io/bitnami/redis-cluster
        depends_on:
          - redis-node-0
          - redis-node-1
          - redis-node-2
          - redis-node-3
          - redis-node-4
        volumes:
            - ./certs:/certs
        environment:
          # - REDIS_TLS_ENABLED=yes
          # - REDIS_PORT_NUMBER=0
          # - REDIS_TLS_PORT=6379
          # - REDIS_TLS_CERT_FILE=/certs/redis.crt
          # - REDIS_TLS_KEY_FILE=/certs/redis.key
          # - REDIS_TLS_CA_FILE=/certs/ca.crt
          - ALLOW_EMPTY_PASSWORD=yes
          - REDIS_CLUSTER_REPLICAS=1
          - REDIS_NODES=redis-node-0 redis-node-1 redis-node-2 redis-node-3 redis-node-4 redis-node-5
          - REDIS_CLUSTER_CREATOR=yes
          - REDIS_CLUSTER_SLEEP_BEFORE_DNS_LOOKUP=5
        networks:
            - nextcloud
```
</details>

<details>
<summary>Docker-compose password protected cluster</summary>

```yml
    redis-node-0:
        image: docker.io/bitnami/redis-cluster
        environment:
          - REDIS_PASSWORD=password
          - REDIS_NODES=redis-node-0 redis-node-1 redis-node-2 redis-node-3 redis-node-4 redis-node-5
        networks:
            - nextcloud

    redis-node-1:
        image: docker.io/bitnami/redis-cluster
        environment:
          - REDIS_PASSWORD=password
          - REDIS_NODES=redis-node-0 redis-node-1 redis-node-2 redis-node-3 redis-node-4 redis-node-5
        networks:
            - nextcloud

    redis-node-2:
        image: docker.io/bitnami/redis-cluster
        environment:
          - REDIS_PASSWORD=password
          - REDIS_NODES=redis-node-0 redis-node-1 redis-node-2 redis-node-3 redis-node-4 redis-node-5
        networks:
            - nextcloud

    redis-node-3:
        image: docker.io/bitnami/redis-cluster
        environment:
          - REDIS_PASSWORD=password
          - REDIS_NODES=redis-node-0 redis-node-1 redis-node-2 redis-node-3 redis-node-4 redis-node-5
        networks:
            - nextcloud

    redis-node-4:
        image: docker.io/bitnami/redis-cluster
        environment:
          - REDIS_PASSWORD=password
          - REDIS_NODES=redis-node-0 redis-node-1 redis-node-2 redis-node-3 redis-node-4 redis-node-5
        networks:
            - nextcloud

    redis-node-5:
        image: docker.io/bitnami/redis-cluster
        depends_on:
          - redis-node-0
          - redis-node-1
          - redis-node-2
          - redis-node-3
          - redis-node-4
        volumes:
            - ./certs:/certs
        environment:
          # - REDIS_TLS_ENABLED=yes
          # - REDIS_PORT_NUMBER=0
          # - REDIS_TLS_PORT=6379
          # - REDIS_TLS_CERT_FILE=/certs/redis.crt
          # - REDIS_TLS_KEY_FILE=/certs/redis.key
          # - REDIS_TLS_CA_FILE=/certs/ca.crt
          - REDIS_PASSWORD=password
          - REDISCLI_AUTH=password
          - REDIS_CLUSTER_REPLICAS=1
          - REDIS_NODES=redis-node-0 redis-node-1 redis-node-2 redis-node-3 redis-node-4 redis-node-5
          - REDIS_CLUSTER_CREATOR=yes
          - REDIS_CLUSTER_SLEEP_BEFORE_DNS_LOOKUP=5
        networks:
            - nextcloud
```
</details>

**Comment by skjnldsv** [comment]
Finally but not last: TLS CLUSTER PASSWORD PROTECTED

```php
    'memcache.locking' => '\OC\Memcache\Redis',
    'memcache.distributed' => '\OC\Memcache\Redis',
    'memcache.local' =>'\OC\Memcache\Redis' ,
    'redis.cluster' => array(
      'seeds' => [
        'redis-node-0:6379',
        'redis-node-1:6379',
        'redis-node-2:6379',
        'redis-node-3:6379',
        'redis-node-4:6379',
        'redis-node-5:6379',
      ],
      'failover_mode' => \RedisCluster::FAILOVER_ERROR,
      'password' => 'password',
      'ssl_context' => [
        'local_cert' => '/certs/redis.crt',
        'local_pk' => '/certs/redis.key',
        'cafile' => '/certs/ca.crt',
        'verify_peer_name' => false
      ]
    ),
```

<details>
<summary>Docker compose config</summary>

```yml
    redis-node-0:
        image: docker.io/bitnami/redis-cluster
        volumes:
            - ./certs:/certs
        environment:
          - REDIS_PORT_NUMBER=0
          - REDIS_TLS_ENABLED=yes
          - REDIS_TLS_AUTH_CLIENTS=yes
          - REDIS_TLS_PORT=6379
          - REDIS_TLS_CERT_FILE=/certs/redis.crt
          - REDIS_TLS_KEY_FILE=/certs/redis.key
          - REDIS_TLS_CA_FILE=/certs/ca.crt
          - REDIS_PASSWORD=password
          - REDIS_NODES=redis-node-0 redis-node-1 redis-node-2 redis-node-3 redis-node-4 redis-node-5
        networks:
            - nextcloud

    redis-node-1:
        image: docker.io/bitnami/redis-cluster
        volumes:
            - ./certs:/certs
        environment:
          - REDIS_PORT_NUMBER=0
          - REDIS_TLS_ENABLED=yes
          - REDIS_TLS_AUTH_CLIENTS=yes
          - REDIS_TLS_PORT=6379
          - REDIS_TLS_CERT_FILE=/certs/redis.crt
          - REDIS_TLS_KEY_FILE=/certs/redis.key
          - REDIS_TLS_CA_FILE=/certs/ca.crt
          - REDIS_PASSWORD=password
          - REDIS_NODES=redis-node-0 redis-node-1 redis-node-2 redis-node-3 redis-node-4 redis-node-5
        networks:
            - nextcloud

    redis-node-2:
        image: docker.io/bitnami/redis-cluster
        volumes:
            - ./certs:/certs
        environment:
          - REDIS_PORT_NUMBER=0
          - REDIS_TLS_ENABLED=yes
          - REDIS_TLS_AUTH_CLIENTS=yes
          - REDIS_TLS_PORT=6379
          - REDIS_TLS_CERT_FILE=/certs/redis.crt
          - REDIS_TLS_KEY_FILE=/certs/redis.key
          - REDIS_TLS_CA_FILE=/certs/ca.crt
          - REDIS_PASSWORD=password
          - REDIS_NODES=redis-node-0 redis-node-1 redis-node-2 redis-node-3 redis-node-4 redis-node-5
        networks:
            - nextcloud

    redis-node-3:
        image: docker.io/bitnami/redis-cluster
        volumes:
            - ./certs:/certs
        environment:
          - REDIS_PORT_NUMBER=0
          - REDIS_TLS_ENABLED=yes
          - REDIS_TLS_AUTH_CLIENTS=yes
          - REDIS_TLS_PORT=6379
          - REDIS_TLS_CERT_FILE=/certs/redis.crt
          - REDIS_TLS_KEY_FILE=/certs/redis.key
          - REDIS_TLS_CA_FILE=/certs/ca.crt
          - REDIS_PASSWORD=password
          - REDIS_NODES=redis-node-0 redis-node-1 redis-node-2 redis-node-3 redis-node-4 redis-node-5
        networks:
            - nextcloud

    redis-node-4:
        image: docker.io/bitnami/redis-cluster
        volumes:
            - ./certs:/certs
        environment:
          - REDIS_PORT_NUMBER=0
          - REDIS_TLS_ENABLED=yes
          - REDIS_TLS_AUTH_CLIENTS=yes
          - REDIS_TLS_PORT=6379
          - REDIS_TLS_CERT_FILE=/certs/redis.crt
          - REDIS_TLS_KEY_FILE=/certs/redis.key
          - REDIS_TLS_CA_FILE=/certs/ca.crt
          - REDIS_PASSWORD=password
          - REDIS_NODES=redis-node-0 redis-node-1 redis-node-2 redis-node-3 redis-node-4 redis-node-5
        networks:
            - nextcloud

    redis-node-5:
        image: docker.io/bitnami/redis-cluster
        depends_on:
          - redis-node-0
          - redis-node-1
          - redis-node-2
          - redis-node-3
          - redis-node-4
        volumes:
            - ./certs:/certs
        environment:
          - REDIS_PORT_NUMBER=0
          - REDIS_TLS_ENABLED=yes
          - REDIS_TLS_AUTH_CLIENTS=yes
          - REDIS_TLS_PORT=6379
          - REDIS_TLS_CERT_FILE=/certs/redis.crt
          - REDIS_TLS_KEY_FILE=/certs/redis.key
          - REDIS_TLS_CA_FILE=/certs/ca.crt
          - REDIS_PASSWORD=password
          - REDISCLI_AUTH=password
          - REDIS_CLUSTER_REPLICAS=1
          - REDIS_NODES=redis-node-0 redis-node-1 redis-node-2 redis-node-3 redis-node-4 redis-node-5
          - REDIS_CLUSTER_CREATOR=yes
          - REDIS_CLUSTER_SLEEP_BEFORE_DNS_LOOKUP=5
        networks:
            - nextcloud
```

</details>

**Comment by skjnldsv** [comment]
Just for info, all my local cluster tests were successful. :rocket: :tada:

**Review by juliusknorr — APPROVED** [review]

**Comment by juliushaertl** [comment]
Hm failures seem related:

```
+ ./occ maintenance:install --admin-pass=admin --data-dir=/dev/shm/nc_int
Nextcloud was successfully installed
+ ./occ config:system:set redis host --value=cache
System config value redis => host set to string cache
+ ./occ config:system:set redis port --value=6379 --type=integer
System config value redis => port set to integer 6379
+ ./occ config:system:set redis timeout --value=0 --type=integer
System config value redis => timeout set to integer 0
+ ./occ config:system:set --type string --value "\OC\Memcache\Redis" memcache.local
System config value memcache.local set to string \OC\Memcache\Redis
+ ./occ config:system:set --type string --value "\OC\Memcache\Redis" memcache.distributed
PHP Warning:  Redis::connect() expects at most 6 parameters, 7 given in /drone/src/lib/private/RedisFactory.php on line 118
An unhandled exception has been thrown:
RedisException: Redis server went away in /drone/src/lib/private/Memcache/Redis.php:55
Stack trace:
#0 /drone/src/lib/private/Memcache/Redis.php(55): Redis->get('63dfb9db467bd4b...')
#1 /drone/src/lib/private/App/InfoParser.php(58): OC\Memcache\Redis->get('/drone/src/apps...')
#2 /drone/src/lib/private/App/AppManager.php(502): OC\App\InfoParser->parse('/drone/src/apps...')
#3 /drone/src/lib/private/legacy/OC_App.php(593): OC\App\AppManager->getAppInfo('files', false, NULL)
#4 /drone/src/lib/private/AppFramework/App.php(69): OC_App::getAppInfo('files')
#5 /drone/src/lib/private/legacy/OC_App.php(278): OC\AppFramework\App::buildAppNamespace('files')
#6 /drone/src/lib/private/AppFramework/Bootstrap/Coordinator.php(108): OC_App::registerAutoloading('files', '/drone/src/apps...')
#7 /drone/src/lib/private/AppFramework/Bootstrap/Coordinator.php(82): OC\AppFramework\Bootstrap\Coordinator->registerApps(Array)
#8 /drone/src/lib/base.php(640): OC\AppFramework\Bootstrap\Coordinator->runInitialRegistration()
#9 /drone/src/lib/base.php(1083): OC::init()
#10 /drone/src/console.php(48): require_once('/drone/src/lib/...')
#11 /drone/src/occ(11): require_once('/drone/src/cons...')
```

**Comment by skjnldsv** [comment]
Fixed failing integration tests

Now this one seems related too

```
There were 4 errors:
157 |  
158 | 1) Test\MemoryInfoTest::testMemoryLimit with data set #0 ('0', 0)
159 | Failed to set memory limit to 0 bytes (Current memory usage is 306184192 bytes)
160 |  
161 | /drone/src/tests/lib/MemoryInfoTest.php:79
162 |  
163 | 2) Test\MemoryInfoTest::testMemoryLimit with data set "134217728 bytes" ('134217728', 134217728)
164 | Failed to set memory limit to 134217728 bytes (Current memory usage is 306184192 bytes)
165 |  
166 | /drone/src/tests/lib/MemoryInfoTest.php:79
167 |  
168 | 3) Test\MemoryInfoTest::testMemoryLimit with data set "128M" ('128M', 134217728)
169 | Failed to set memory limit to 134217728 bytes (Current memory usage is 306184192 bytes)
170 |  
171 | /drone/src/tests/lib/MemoryInfoTest.php:79
172 |  
173 | 4) Test\MemoryInfoTest::testMemoryLimit with data set "131072K" ('131072K', 134217728)
174 | Failed to set memory limit to 134217728 bytes (Current memory usage is 306184192 bytes)
175 |  
176 | /drone/src/tests/lib/MemoryInfoTest.php:79
177 |  
178

```

**Comment by juliushaertl** [comment]
Looks related to https://php.watch/versions/7.4/memory_limit-lt-2m

I think we would need to adjust the tests so that the memory consumption of running the whole test suite would always fit in the tested values. 

In addition https://github.com/nextcloud/server/blob/2a529e453a7de7f51ab72ac48f947fa9d2afd18d/tests/lib/MemoryInfoTest.php#L63 would need to get dropped, but setting the memory limit to 0 doesn't make sense anyways :wink:

**Comment by juliushaertl** [comment]
Most likely failing in the tests even though it is a warning due to the default in PHPUnit:

> The convertWarningsToExceptions Attribute
> 
> Possible values: true or false (default: true)
> 
> This attribute configures whether E_WARNING and E_USER_WARNING events triggered by the code under test are converted to an exception (and mark the test as error).

**Comment by PVince81** [comment]
this introduced a little regression for those like me who relied on default values, PR here to fix it: https://github.com/nextcloud/server/pull/28129

**Pull request merged**